GiniVision/Classes/ToolTipView.swift

```diff
@@ -0,0 +1,303 @@
+        self.textSize = size(forText: text, withFont: font)
+        self.addTipContainer(backgroundColor: backgroundColor)
+        self.addTextLabel(withText: text, textColor: textColor, font: font)
+        self.addCloseButton(withColor:closeButtonColor)
+        self.addArrow()
+        self.addShadow()
+        
+        self.arrangeFrame(withSuperView: superView)
+        self.arrangeArrow(withSuperView: superView)
+        self.setupConstraints()
+    }
+    
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        self.arrangeArrow(withSuperView: superview)
+    }
+    
+    func arrangeViews() {
+        self.arrangeFrame(withSuperView: superview)
+        self.arrangeArrow(withSuperView: superview)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("NSCoding not supported. Use init(text) instead!")
+    }
+    
+    // MARK: Add views
+    fileprivate func addTipContainer(backgroundColor color:UIColor) {
+        self.addSubview(tipContainer)
+        self.tipContainer.backgroundColor = color
+    }
+    
+    fileprivate func addTextLabel(withText text:String, textColor:UIColor, font:UIFont) {
+        textLabel.text = text
+        textLabel.textColor = textColor
+        textLabel.font = font
+        textLabel.numberOfLines = 0
+        tipContainer.addSubview(textLabel)
+    }
+    
+    fileprivate func addArrow() {
+        arrowView.frame.origin = CGPoint(x: 0, y: self.frame.height)
+        self.addSubview(arrowView)
+    }
+    
+    fileprivate func addCloseButton(withColor color:UIColor) {
+        let image = UIImageNamedPreferred(named: "toolTipCloseButton")
+        closeButton.setImage(image, for: .normal)
+        closeButton.tintColor = color
+        closeButton.addTarget(self, action: #selector(closeAction), for: .touchUpInside)
+        tipContainer.addSubview(closeButton)
+    }
+    
+    fileprivate func addShadow() {
+        self.layer.shadowOffset = CGSize(width: 0, height: 2)
+        self.layer.shadowRadius = 0.8
+        self.layer.shadowOpacity = 0.2
+        self.layer.shadowColor = UIColor.black.cgColor
+    }
+    
+    // MARK: Arrange views
+    fileprivate func arrangeFrame(withSuperView superView:UIView?) {
+        guard let superview = superView, let referenceViewAbsoluteFrame = absoluteFrame(for: referenceView, inside: superView) else { return }
+        let frameHeight = max(textSize.height, closeButtonHeight) + padding.top + padding.bottom + margin.top + margin.bottom
+        let frameWidth = textSize.width + closeButtonWidth + padding.left + padding.right + itemSeparation + margin.left + margin.right
+        let size = CGSize(width: frameWidth, height: frameHeight)
+        
+        var x:CGFloat = 0
+        var y:CGFloat = 0
+        switch toolTipPosition {
+        case .above:
+            if referenceViewAbsoluteFrame.origin.y - size.height < 0 {
+                assertionFailure("The tip cannot be shown outside the super view")
```

Comment:
Could these assertions lead to crashing the app? Can these be caught? Would be bad if our lib crashes apps when for some reason the super view's size is wrong.

As an alternative we could fail gracefully and use alternative positions if the desired one cannot be used.

Reply:
It should only crash to us when we are developing since this class is not public and we control where the tip should appear.
The worst case for the company is that they have the text truncated if they embed the cameraViewController inside a small view (quite small).